kong/clustering/rpc/concentrator.lua

```diff
@@ -0,0 +1,299 @@
+      local n, err = notifications_queue:pop(0)
+      if err then
+        return nil, "unable to pop from notifications queue: " .. err
+      end
+
+      if n then
+        assert(n.operation == "notification")
+
+        if n.channel == rpc_resp_channel_name then
+          -- an response for a previous RPC call we asked for
+          local payload = assert(cjson_decode(n.payload))
+          assert(payload.jsonrpc == "2.0")
+
+          -- response
+          local cb = self.interest[payload.id]
+          self.interest[payload.id] = nil -- edge trigger only once
+
+          if cb then
+            local res, err = cb(payload)
+            if not res then
+              ngx_log(ngx_WARN, "[rpc] concentrator response interest handler failed: id: ",
+                      payload.id, ", err: ", err)
+            end
+
+          else
+            ngx_log(ngx_WARN, "[rpc] no interest for concentrator response id: ", payload.id, ", dropping it")
+          end
+
+        else
+          -- other CP inside the cluster asked us to forward a call
+          assert(n.channel:sub(1, #REQ_CHANNEL_PREFIX) == REQ_CHANNEL_PREFIX)
+
+          local target_id = n.channel:sub(9)
+          local sql = string_format(RPC_REQUEST_DEQUEUE_SQL, self.db.connector:escape_literal(target_id))
+          local calls, err = self.db.connector:query(sql)
+          if not calls then
+            return nil, "concentrator request dequeue query failed: " .. err
+          end
+
+          assert(calls[1] == true)
+          ngx_log(ngx_DEBUG, "concentrator got ", calls[2].affected_rows,
+                  " calls from database for node ", target_id)
+          for _, call in ipairs(calls[2]) do
+            local payload = assert(cjson_decode(call.payload))
+            local reply_to = assert(call.reply_to,
+                                    "unknown requester for RPC")
+
+            local res, err = self.manager:_call(target_id, payload.method,
+                                                payload.params)
+            if res then
+              -- call success
+              res, err = self:_enqueue_rpc_response(reply_to, {
+                jsonrpc = "2.0",
+                id = payload.id,
+                result = res,
+              })
+              if not res then
+                ngx_log(ngx_WARN, "[rpc] unable to enqueue RPC call result: ", err)
+              end
+
+            else
+              -- call failure
+              res, err = self:_enqueue_rpc_response(reply_to, {
+                jsonrpc = "2.0",
+                id = payload.id,
+                ["error"] = {
+                  code = jsonrpc.SERVER_ERROR,
+                  message = tostring(err),
+                }
+              })
+              if not res then
+                ngx_log(ngx_WARN, "[rpc] unable to enqueue RPC error: ", err)
+              end
+            end
+          end
+        end
+      end
+    until not n
+
+    local res, err = lconn:wait_for_notification()
```

Comment:
The way concentrator uses Postgres is very special, it does not work with the usual connector pooling logic. Therefore, we can not use the standard connector because this will break other code calling `kong.db.xxx` directly or worse, lose some Postgres notifications accidentally. Interacting with low-level `pgmoon` instance is the safe thing to do due to how special this use case is.